sinks/xray/xray.go

```diff
@@ -0,0 +1,228 @@
+type XRaySpanSink struct {
+	daemonAddr      string
+	traceClient     *trace.Client
+	conn            *net.UDPConn
+	sampleThreshold uint32
+	commonTags      map[string]string
+	log             *logrus.Logger
+	spansDropped    int64
+	spansHandled    int64
+	nameRegex       *regexp.Regexp
+}
+
+var _ sinks.SpanSink = &XRaySpanSink{}
+
+// NewXRaySpanSink creates a new instance of a XRaySpanSink.
+func NewXRaySpanSink(daemonAddr string, sampleRatePercentage int, commonTags map[string]string, log *logrus.Logger) (*XRaySpanSink, error) {
+
+	log.WithFields(logrus.Fields{
+		"Address": daemonAddr,
+	}).Info("Creating X-Ray client")
+
+	var sampleThreshold uint32
+	if sampleRatePercentage < 0 {
+		log.WithField("sampleRatePercentage", sampleRatePercentage).Warn("Sample rate < 0 is invalid, defaulting to 0")
+		sampleRatePercentage = 0
+	}
+	if sampleRatePercentage > 100 {
+		log.WithField("sampleRatePercentage", sampleRatePercentage).Warn("Sample rate > 100 is invalid, defaulting to 100")
+		sampleRatePercentage = 100
+	}
+
+	// Set the sample threshold to (sample rate) * (maximum value of uint32), so that
+	// we can store it as a uint32 instead of a float64 and compare apples-to-apples
+	// with the output of our hashing algorithm.
+	sampleThreshold = uint32(sampleRatePercentage * math.MaxUint32 / 100)
+
+	// Build a regex for cleaning names based on valid characters from:
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+	reg, err := regexp.Compile(`[^a-zA-Z0-9_\.\:\/\%\&#=+\-\@\s\\]+`)
+	if err != nil {
+		return nil, err
+	}
+
+	return &XRaySpanSink{
+		daemonAddr:      daemonAddr,
+		sampleThreshold: sampleThreshold,
+		commonTags:      commonTags,
+		log:             log,
+		nameRegex:       reg,
+	}, nil
+}
+
+// Start the sink
+func (x *XRaySpanSink) Start(cl *trace.Client) error {
+	x.traceClient = cl
+
+	xrayDaemon, err := net.ResolveUDPAddr("udp", x.daemonAddr)
+	if err != nil {
+		return err
+	}
+	conn, err := net.DialUDP("udp", nil, xrayDaemon)
+	if err != nil {
+		return err
+	}
+	x.conn = conn
+
+	return nil
+}
+
+// Name returns this sink's name.
+func (x *XRaySpanSink) Name() string {
+	return "xray"
+}
+
+// Ingest takes in a span and passed it along to the X-Ray client after
+// some sanity checks and improvements are made.
+func (x *XRaySpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+
+	sampleCheckValue := []byte(strconv.FormatInt(ssfSpan.TraceId, 10))
+	hashKey := crc32.ChecksumIEEE(sampleCheckValue)
+	if hashKey > x.sampleThreshold {
+		atomic.AddInt64(&x.spansDropped, 1)
+		return nil
+	}
+
+	metadata := map[string]string{}
+	for k, v := range x.commonTags {
+		metadata[k] = v
+	}
+	for k, v := range ssfSpan.Tags {
+		if k != XRayTagNameClientIP {
+			metadata[k] = v
+		}
+	}
+	if ssfSpan.Indicator {
+		metadata["indicator"] = "true"
+	} else {
+		metadata["indicator"] = "false"
```

Comment:
:lurking: heads up that this hash never gets attached to the XRaySegment object

Reply:
it does now!